contracts/TokenConverter/AbstractTokenConverter.sol

```diff
@@ -683,13 +807,92 @@ abstract contract AbstractTokenConverter is AccessControlledV8, IAbstractTokenCo
+        uint256 convertedTokenInBalance;
+        if (address(converterNetwork) != address(0)) {
+            (address[] memory converterAddresses, uint256[] memory converterBalances) = converterNetwork
+            .findTokenConvertersForConverters(tokenAddressOut, tokenAddressIn);
+            uint256 convertersLength = converterAddresses.length;
+            for (uint256 i; i < convertersLength; ) {
+                if (converterBalances[i] == 0) break;
+                (, uint256 amountIn) = IAbstractTokenConverter(converterAddresses[i]).getUpdatedAmountIn(
+                    converterBalances[i],
+                    tokenAddressOut,
+                    tokenAddressIn
+                );
+                if (amountIn > amountToConvert) {
+                    amountIn = amountToConvert;
+                }
+
+                uint256 balanceBefore = IERC20Upgradeable(tokenAddressIn).balanceOf(_destinationAddress);
+
+                IERC20Upgradeable(tokenAddressOut).approve(converterAddresses[i], amountIn);
+                IAbstractTokenConverter(converterAddresses[i]).convertExactTokens(
+                    amountIn,
+                    0,
+                    tokenAddressOut,
+                    tokenAddressIn,
+                    _destinationAddress
+                );
+
+                uint256 balanceAfter = IERC20Upgradeable(tokenAddressIn).balanceOf(_destinationAddress);
+                amountToConvert -= amountIn;
+                convertedTokenInBalance += (balanceAfter - balanceBefore);
+
+                if (amountToConvert == 0) break;
+                unchecked {
+                    ++i;
+                }
+            }
+        }
+
+        _postPrivateConversion(comptroller, tokenAddressIn, convertedTokenInBalance, tokenAddressOut, amountToConvert);
+    }
+
+    /// @dev This hook is used to update states for the converter after the privateConversion
+    /// @param comptroller Comptroller address (pool)
+    /// @param tokenAddressIn Address of the destination's base asset
+    /// @param convertedTokenInBalance Amount of the base asset received after the conversion
+    /// @param tokenAddressOut Address of the asset transferred to other converter in exchange of base asset
+    /// @param convertedTokenOutBalance Amount of tokenAddressOut transferred from this converter
+    function _postPrivateConversion(
+        address comptroller,
+        address tokenAddressIn,
+        uint256 convertedTokenInBalance,
+        address tokenAddressOut,
+        uint256 convertedTokenOutBalance
+    ) internal virtual {}
+
+    /// @notice This hook is used to update the state for asset reserves before transferring tokenOut to user
+    /// @param tokenOutAddress Address of the asset to be transferred to the user
+    /// @param amountOut Amount of tokenAddressOut transferred from this converter
+    function preTransferHook(address tokenOutAddress, uint256 amountOut) internal virtual {}
```

Comment:
We should use a single naming convention for hooks IMO, either `_preTransfer` and `_postPrivateConversion` or `preTransferHook` and `postPrivateConversionHook`, or the latter with an underscore. I don't have a preference here, any unified approach is good.

(I personally like all internal functions to start with `_` because I came from C++/Python world, but in Solidity there's no such strict practice I think)